### PR TITLE
Add in local variable completion and more complete AST

### PR DESCRIPTION
**Pull request by seamusriordan**
This PR adds in local variable completion to complement the existing global define completion.  This should handle forms from `def`, `defn(-)`, `defmacro(-)`, lambdas, `let`, `for`, `each`, and `loop` with correct scoping.  I included automated testing for as many unique cases as I could consider and it behaves well on its own source.  I did the best I could, but I'm just one person so any more manual testing is probably a good idea.

I modified and extended the existing PEG from `fmt` to suit this but left the one in `fmt` alone so as not to break it.  I imagine there's some room for refactoring to consolidate those.  This version now picks out where parameters and variables get defined, includes their positions in the text, and does the same for the symbols in `:span`.  I switched to using structs so getting data by keywords is straightforward.

I did find an edge case in word-at for words at the end of a line or `]`, `}`, and `"` (the `)` case had been considered).  I updated those tests accordingly (you can see the difference in range now matches the expected word length). 

The AST this produces should include all symbol declarations as well as their uses with the positions.  This I believe should put it in a good spot to be able to do the LSP renaming, go-to-declaration, and find usage functionality by walking through the tree appropriately.

There were some tests which did not pass out-of-the-box which I ignored (some may be local to a different development environment?) as well as some which have commit hashes (I presume there's some build process to get those).  The remaining tests are still passing on my end, at least as far as `judge` claims.

### CODE REVIEW

File: src/lookup.janet [hunk 1]
```diff
@@ -13,9 +13,9 @@
              ,|[$0 $1 (dec $2)])
       :word (/ (* (position)
                   (<- (some (if-not (set " ()[]{}`'\"") 1)))
-                  (? (+ :s ")"))
-                  (position))
-               ,|[$0 $1 (dec $2)])
+                  (position)
+                  (? (+ :s (set ")}]\""))))
```
Comment (seamusriordan) [inline src/lookup.janet:17]: This should catch the "terminating" cases and instead of decrementing uses another `position` where the word ends.

**Comment by CFiggers** [comment]
Wow, awesome!! I've wanted this feature for a long time.

Initial impressions:

* Really interesting bug with `word-at`—I reckon it's somewhat uncommon in my own code that I'm looking for a hover pop-up on a symbol that either a) is inside and at the end of a string (the `"` case), b) is the final value of a dictionary (the `}` case) or c) is the final value in a list (the `]` case). So that would be why I hadn't noticed that before now. But you're totally right, that's broken in the current master and needs to be fixed!
  * **I kinda want to handle that as a separate pull request just for this bugfix.** Would you be interested in packaging that up and submitting separately? If not I can handle that.

* Good call on replicating the peg from `libs/fmt` to make your extensions and modifications—that module should stay in sync with `spork/fmt`, which is where that code is vendored from.

* Thank you SO MUCH for the extensive tests. That makes review much easier and would be something I would ask for if you didn't beat me to it.

* LOVE what you say about potentially being able to reuse this same AST mechanism to implement `textDocument/rename` and `textDocument/references`.

This is super exciting. I really appreciate the high effort PR. I'm going to do a deep dive this weekend and we'll see how quickly we can get it merged. If I find things I'd like to update, would you prefer that I leave comments and you can make your own edits, or should I commit directly where called for?

Thank you so much!

**Comment by CFiggers** [comment]
Thank you again for your work on this! This feature is _so cool_ and I'm really enjoying it already!

Some explanations of the commits I added:

* Move `utils.janet` to `src/`
  * `libs/` is reserved for code that is vendored from third-party sources (`fmt.janet` from `janet-lang/spork`, `jayson.janet` from `cfiggers/jayson`) or supporting code that's tightly coupled to external libraries (`jpm-defs.janet` relating to `janet-lang/jpm`).
* Use `varfn` instead of `var` and `set`
  * I noticed that you were using `var` and `set` at the top level in `parser.janet` to make advance declarations of functions in order to enable recursive dependencies between `get-syms-from-tree`, `find-symbols-in-nodes`, and `find-symbol-in-node`. 
  * Janet has `varfn` in the core library for this purpose. 
  * Also, it's only necessary to have one (1) function be a `varfn`, the other two can be standard immutable `defn`s.
* Use `catseq` instead of custom util `flatmap`
  * I noticed that your `flatmap` function in `utils.janet` essentially just `map`s a function over a data structure and then calls `flatten` on the result.
  * Janet has `catseq` in the core library for this purpose. (Technically your version is more flexible because it could theoretically take multiple `vs` to `map` over. But in practice we seemed to only be using `flatmap` with a single data structure, so the `catseq`s I substituted in _shouldn't_ encounter any issues, I don't think.)
    * Also, I'm not sure precisely what the underlying mechanism is, but the "comprehension" functions in Janet's core library (`seq`, `catseq`, `tabseq`) are _significantly_ faster than `map`. So we should get some improved performance from this change.
  * I also inlined a couple of private `defn-`s you had in `parser.janet`. 
    * Janet, being a Virtual Machine implemented in C, _does_ have a small performance overhead penalty every time we invoke a function. 
    * The compiler is not currently smart enough to identify when you've only used a function a single time. So, it's a tiny performance optimization to have one slightly longer, slightly hairier function rather than split things up into tiny functions that all call one another (it's a close-to-negligible difference in this case, but since `textDocument/completion` and/or `completionItem/resolve` can get called between every keystroke in the client's editor, the faster we _can_ be the better).

Let me know if you have any feedback on or questions about any of the above.

Thanks once again, especially for the extensive Judge tests. I felt very confident refactoring because after every tweak I made all the tests still passed. 

I don't think I would've gotten this implemented myself _any_ time in the near future. I'm going to run the current master branch for a little while to see if any additional bugs or edge cases shake out. But assuming not, I'll do a v0.0.11 release shortly and then ship this update in an update of [Janet++](https://github.com/cfiggers/vscode-janet-plus-plus)!
